lib/filterx/expr-comparison.c

```diff
@@ -44,43 +45,43 @@ typedef struct _FilterXComparison
 static void
 _convert_filterx_object_to_generic_number(FilterXObject *obj, GenericNumber *gn)
 {
-  if (filterx_object_is_type(obj, &FILTERX_TYPE_NAME(integer)) ||
-      filterx_object_is_type(obj, &FILTERX_TYPE_NAME(double)) ||
-      filterx_object_is_type(obj, &FILTERX_TYPE_NAME(boolean)))
-    *gn = filterx_primitive_get_value(obj);
-  else if (filterx_object_is_type(obj, &FILTERX_TYPE_NAME(string)))
+  if (filterx_object_extract_generic_number(obj, gn))
+    return;
+
+  const gchar *str;
+  if (filterx_object_extract_string(obj, &str, NULL))
     {
-      if (!parse_generic_number(filterx_string_get_value(obj, NULL), gn))
+      if (!parse_generic_number(str, gn))
         gn_set_nan(gn);
+      return;
     }
-  else if (filterx_object_is_type(obj, &FILTERX_TYPE_NAME(null)))
-    gn_set_int64(gn, 0);
-  else if (filterx_object_is_type(obj, &FILTERX_TYPE_NAME(datetime)))
+
+  if (filterx_object_extract_null(obj))
     {
-      const UnixTime utime = filterx_datetime_get_value(obj);
-      uint64_t unix_epoch = unix_time_to_unix_epoch(utime);
-      gn_set_int64(gn, (uint64_t)unix_epoch);
+      gn_set_int64(gn, 0);
+      return;
     }
-  else
+
+  UnixTime utime;
+  if (filterx_object_extract_datetime(obj, &utime))
     {
-      gn_set_nan(gn);
+      uint64_t unix_epoch = unix_time_to_unix_epoch(utime);
+      gn_set_int64(gn, (uint64_t) unix_epoch);
```

Comment:
I think you might have wanted to cast this to `gint64`

Reply:
Thanks, fixed.